### PR TITLE
Performance: remove forEach & some unneeded work

### PR DESCRIPTION
**Pull request by STRML**
Part of npm optimization; shows 2-5x payoff in calls to
TrackerGroup.complete()

For comparison:

##### Before this PR:

screen shot 2016-01-27 at 11 26 23 am

##### After:

screen shot 2016-01-27 at 12 22 39 pm

##### Before:

screen shot 2016-01-27 at 11 26 37 am

##### After:

screen shot 2016-01-27 at 12 26 17 pm

Performance difference is actually about 100x in this case, likely due to deopt causing many allocations + GC pressure.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,24 +8,23 @@ var TrackerGroup = exports.TrackerGroup = function (name) {
   EventEmitter.call(this)
   this.name = name
   this.trackGroup = []
-  var self = this
   this.totalWeight = 0
-  var noteChange = this.noteChange = function (name) {
-    self.emit("change", name || this.name)
-  }.bind(this)
```
Comment (iarna) [inline index.js]: O.o I ... uh, I don't know why I was both wrapping `self` AND binding, but that's obviously wrong.

File: index.js [hunk 2]
```diff
@@ -8,24 +8,23 @@ var TrackerGroup = exports.TrackerGroup = function (name) {
...
-  this.trackGroup.forEach(function(unit) {
-    unit.on("change", noteChange)
-  })
+  this.noteChange = TrackerGroup.prototype.noteChange.bind(this);
```
Comment (iarna) [inline index.js]: This is actually slower than changing the above `this.name` to `self.name` and ditching the bind, because bind is terrifyingly slow.

Comment (STRML) [inline index.js]: It's actually [not anymore](https://chromium.googlesource.com/v8/v8/+/97def8070cbb979419c041c3732d5e11779f79da) (I don't think this has hit Node yet), but I agree, could just make it a closure so it's fast on older engines too.

**Comment by iarna** [comment]
I've merged this as 48fd51d, f5fa25d & dab2fed (with a little tweaking).

**Comment by STRML** [comment]
Great - thanks.

**Pull request closed without merging**